### DISCUSSION TITLE
How to run Prisma studio locally and connect to remote server

### DISCUSSION THREAD

**Discussion by jcubic** [prisma/prisma#18537] [Q&A]
I've edited `.env` file and put my remote host, and the reset of the credential are the same (I've created the same connection locally as created by my shared hosting).

The problem is that when I run Prisma studio and I got a local database instead of a remote one.

**Comment by ludralph** [comment]
Hi @jcubic 👋 

Can you please share the content of the `.env` file? Please kindly remove sensitive data before sharing.

**Reply by jcubic** [reply]
```
PORT=3000
DEBUG=1

SMTP_HOST=<HOST>
SMTP_USER=<EMAIL>
SMTP_PASS=<PASSWORD>

ADMIN_NAME=<USER>
ADMIN_PASS=<PASS>

DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD>@localhost:3306/10669_koduj_quiz"

SESSION_SECRET="<SECRET>"
```

I've changed localhost to remote server and I when run `npx prisma studio` I still get a local database.

**Reply by ludralph** [reply]
Please confirm that the environment variable names in your .env file match those used in your Prisma client configuration file. Also, did you restart your app after making changes to the `.env` file?

**Reply by jcubic** [reply]
Do you talk about `prisma/schema.prisma` I don't have any other files.

```
datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}
```

Why do I need to restart my app, this for prisma studio, my app is not even running.
But I just tried to run it after changing `.env` and then running studio again and have the same issue, it read localost database but the same as my app, it looks like host in the URL is ignored.

**Reply by janpio** [reply]
What do you get when you execute `echo $DATABASE_URL` in your terminal where you also execute Prisma Studio?

**Reply by jcubic** [reply]
@janpio Nothing. I don't have that variable defined in the terminal.

**Reply by janpio** [reply]
Then something is off. Prisma Studio reads either from the `.env` file and puts the values into the environment, or it uses values that are already in the environment.

What happens when you remove `DATABASE_URL` from `.env` completely? Does it still connect to a database? Where could it get that value from? (Studio is not persisted to disk at all, but fully reloaded when you run `npx prisma studio`)

**Reply by jcubic** [reply]
@janpio got error like this:

```
  55 
  56 app.get('/quiz/:id/:slug?', is_auth, async function(req: Request, res: Response) {
  57     const poll_id = +req.params.id;
→ 58     const poll = await prisma.poll.findFirst(
error: Environment variable not found: DATABASE_URL.
  -->  schema.prisma:8
   | 
 7 |   provider = "mysql"
 8 |   url      = env("DATABASE_URL")
   |
```

Maybe the code doesn't use the host and always uses localhost, I was looking at the unit tests and I've only seen localhost. The only different host is for tests that throw an error. Are you sure that you can connect prisma to a different host? I know this is kind of silly but maybe this is what is happening.

**Reply by janpio** [reply]
Yes, of course. Prisma connects to the URL that is defined in `url`, and if that is via `env("DATABASE_URL")` then it will read the value of that env var. If a `.env` file is present it will read that into the environment first.

**Reply by jcubic** [reply]
@janpio I was asking if it reads the HOST because it seems that it uses localhost as default.

**Reply by janpio** [reply]
I don't understand. If there is a connection string, it includes a hostname (usually after the protocol and credentials, and after a slash that then leads over to the database name). Prisma uses whatever is in that part of the connection string and tries to connect to it. What does your `DATABASE_URL` value look like?

**Reply by jcubic** [reply]
From my experience with Prisma, it always uses localhost.

This is what I have:
```
DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD>@jcubic.atthost24.pl:3306/10669_koduj_quiz"
```

and the host is ignored, I'm not able to see my updated data. I use the same credentials for both remote and local databases and it simply doesn't work. And the code is hard to read and hard to find what Prisma is actually doing.

To be more precise, I have a quiz and locally answered one question on the server answered all. And when running prisma studio it shows one answer. I'm not even able to edit the data on the server because of this.

**Reply by janpio** [reply]
I can only repeat: It uses exactly the host defined in the URL, in this case `jcubic.atthost24.pl` - assuming this `DATABASE_URL` is correctly read into your environment variables and there is not already another value present.

A simple way to make sure nothing can go wrong there is to just set different credentials on the remote and the local database. Then you will get an error message when the wrong credentials are used to connect to a specific host.

**Reply by janpio** [reply]
What happens if you manually set the environment variable to this value via `export DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD>@jcubic.atthost24.pl:3306/10669_koduj_quiz"` before running your script?

**Reply by jcubic** [reply]
@janpio I changed the password on the server, got an error changed in the DATABASE_URL got my server data. Changed the password back to old password in both places and I have localhost.

The only conclusion is that Prisma use some kind of cache, maybe store the credentials with some kind of hash, and uses whatever host it was assigned to that user/password. This looks like a bug to me.

**Reply by janpio** [reply]
There is no cache. Prisma reads the value of `DATABASE_URL` from the environment when you run the script or CLI. Really.

> I changed the password on the server, got an error changed in the DATABASE_URL got my server data. Changed the password back to old password in both places and I have localhost.

I am unfortunately not completely following. Can you explain more precisely here? 
What connection string did you have in place before changing anything? 
What did you do exactly step by step?

**Reply by jcubic** [reply]
Ok maybe I was not clear enough, even that I've written all the steps:

1. use this in my `.env` `DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD_1>@localhost:3306/10669_koduj_quiz"`
2. Change localhost to my server name `DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD_1>@jcubic.atthost24.pl:3306/10669_koduj_quiz"` - no effect it reads from localhost
3. Change Server password and change .env to `DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD_2>@jcubic.atthost24.pl:3306/10669_koduj_quiz"` - it reads server data
4. Change back Server password and .env to `DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD_1>@jcubic.atthost24.pl:3306/10669_koduj_quiz"` - it reads again local database instead of remote

Maybe you use some library that caches the URL, it may be not your code base that is causing the issue.

**Reply by janpio** [reply]
When you restart your app, any context of Prisma is discarded. No state of the engine is persisted in any way.

Something _has_ to be off here.

Can you reproduce the problem in a new project created with `prisma init` in a new folder? (Use `db pull` on the database to quickly create the schema, then create a minimal `script.js` that reads data in a way that you will be able to confirm if this is happening or not)

**Comment by janpio** [comment]
Additionally, the following steps could help us understand (As soon as something does not happen as described, you can stop and report what happened instead):

1. `.env` = `DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD_1>@localhost:3306/10669_koduj_quiz"`  
Start your server with this configuration.  
Confirm that "localhost" data is read.
2. `.env` = `DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD_1>@jcubic.atthost24.pl:3306/10669_koduj_quiz"`  
Start your server with this configuration.  
Confirm that "localhost" data is read.  
(This is what in my, Prisma's, opinion should already be impossible unless you have routed that host locally to `localhost`)
3. Change password to something else on your local server.
4. Go back to 1), does it throw an error now?
5. Go back to 2), does it throw an error now?
6. Turn off the locally running server,
7. Go back to 1), does it time out now?
8. Go back to 2), does it time out now?

**Reply by jcubic** [reply]
How can I start prisma properly? What do I need to restart? I have ts-node that run the app and I kill it and run again. THe same as with prisma studio, this is normal that I need to restart the app for the configuration to be applied. Do I need to do any extra steps to restart the app? Like generate new Prisma files?

Chaning password locally throws an error (already told you) only if the user and password is the same as before it read the same host even if it changes.

I will check the blank project and get back to you.

**Reply by janpio** [reply]
> How can I start prisma properly? What do I need to restart? I have ts-node that run the app and I kill it and run again. THe same as with prisma studio, this is normal that I need to restart the app for the configuration to be applied. Do I need to do any extra steps to restart the app? Like generate new Prisma files?

You kill the running server/process, and start a new one.
If you configured your `schema.prisma` to read from an environment variable, that will happen again when you start the process again.
So it seems your are doing it correctly already.

> Chaning password locally throws an error (already told you) only if the user and password is the same as before it read the same host even if it changes.

Sorry, as before I am not following this sentence. It is missing some punctuation, or some information, to make sense to me. Hence, optimally answer with the same numbers that I used in the list above and concretely list what happens at each step.

**Reply by jcubic** [reply]
2. `.env` = `DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD_1>@jcubic.atthost24.pl:3306/10669_koduj_quiz"`
3. `.env` = `DATABASE_URL="mysql://10669_koduj_quiz:<PASSWORD_2>@jcubic.atthost24.pl:3306/10669_koduj_quiz"`

it gives an error and reads the server database with the wrong password. So in my case changing a password is the only way to change the host. This makes no sense, but this is what is happening.

**Reply by janpio** [reply]
What does "the server database with the wrong password" mean?

Please really just do the steps from https://github.com/prisma/prisma/discussions/18537#discussioncomment-5512196 and for each of these numbers report exactly what happens as requests.

**Reply by jcubic** [reply]
It means, I change the password in DATABSE_URL environment variable to the password that is invalid for the server but different. Look at the steps, strings are the same but the password is different and there are no step to change password on the server.

**Reply by jcubic** [reply]
Sorry for my last comment, I've checked your steps and:

Step 2 it reads localhost
Step 4 it throws an error
Step 5 it reads the remote server data (I have the same password on the server `PASSWORD_1`).
Step 7 it threw error (it can't connect)
Step 8 it reads the remote server data like in 5 (I still have the same password on the server `PASSWORD_1`).

but in the beginning, I had the same password on both the server and localhost.

Maybe this is something with MySQL database or NodeJS library that connects to MySQL and not with Prisma, just guessing.

I use this on Fedora GNU Linux and I use MariaDB compatible with MySQL.

I will ask on some StackExchage site maybe ServerFault and get back to you.

**Reply by janpio** [reply]
When you use Prisma, there is no Node library for MySQL involved as that is all run via Prisma Engine that uses Rust.

What you are observing is totally unexpected and weird to me I think. I have no idea how this can be happening. I hope someone can point out to you what might be going on here.

**Reply by jcubic** [reply]
This is really odd, even if I stopped the server and run it again I got localhost when using the host. Maybe this is something with my Linux system. As for now, I will just stop the local server, since this will load the remote server.

I've asked on ServerFault: and they suggest that if the connection is open it will be reused. But I kill the app. Is there any background process in Prisma Studio when calling `npx prisma studio`?

[Is it possible that MariaDB caches the host depending on user and password?](https://serverfault.com/q/1128138/54695)

**Reply by janpio** [reply]
No. When you end the `prisma studio` process, everything is shut down. Only when you run the command again, a new instance will be started that starts completely from nothing.

Can you also reproduce the problem with your own app, or only with Prisma Studio?

**Reply by jcubic** [reply]
It seems the same as on a normal application, I have a quiz app and deleted all the answers from the studio, run the app, and it's local version without answers. My hosted app was not touched. And the same happens when I stopped MySQL (mariaDB) server, both the studio and my app read the host database.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
